modules/ipp/scanner.go

```diff
@@ -225,80 +211,149 @@ func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
 		}
 	}
-	// Write reported IPP versions to results object
-	if len(scan.results.AttributeIPPVersions) == 0 {
-		if ippVersions, err := readAttributeFromBody(VersionsSupported, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": VersionsSupported,
-			}).Debug("Failed to read attribute.")
-		} else {
-			for _, v := range ippVersions {
-				scan.results.AttributeIPPVersions = append(scan.results.AttributeIPPVersions, string(v))
+}
+
+// TODO: Log every error that could come out of this
+// TODO: Determine whether errors should be ignored, debug logged, fatal, etc.
+func readAllAttributes(body *[]byte) (*[]*Attribute) {
+	buf := bytes.NewBuffer(*body)
+	var start struct {
+		Version int16
+		StatusCode int16
+		ReqID int32
+	}
+	binary.Read(buf, binary.BigEndian, &start)
+	var tag byte
+	binary.Read(buf, binary.BigEndian, &tag)
+	var groups []AttributeGroup
+	// Until encountering end-of-attributes-tag:
+	// Per-group loop
+	for tag != 0x03 {
+		// Create a new group
+		group := AttributeGroup{Tag: tag}
+		// Read the first tag:
+		binary.Read(buf, binary.BigEndian, &tag)
+		var lastTag byte
+		for tag < 0x00 || tag > 0x05 {
+			// TODO: Implement parsing attribute collections (they're special)
+			var attr *Attribute
+			var nameLength int16
+			binary.Read(buf, binary.BigEndian, &nameLength)
+			if tag == lastTag && nameLength == 0 {
+				attr = group.Attrs[len(group.Attrs)-1]
+			} else {
+				attr = &Attribute{ValueTag: tag}
+				name := make([]byte, nameLength)
+				binary.Read(buf, binary.BigEndian, &name)
+				attr.Name = string(name)
+			}
+			var length int16
+			binary.Read(buf, binary.BigEndian, &length)
+			val := make([]byte, length)
+			binary.Read(buf, binary.BigEndian, &val)
+			attr.Values = append(attr.Values, Value{Bytes: val})
+
+			if len(group.Attrs) == 0 || attr != group.Attrs[len(group.Attrs)-1] {
+				group.Attrs = append(group.Attrs, attr)
 			}
+			lastTag = tag
+			binary.Read(buf, binary.BigEndian, &tag)
 		}
+		groups = append(groups, group)
 	}
-	// Write reported printer URI to results object
-	if scan.results.AttributePrinterURI == "" {
-		if uris, err := readAttributeFromBody(PrinterURISupported, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": PrinterURISupported,
-			}).Debug("Failed to read attribute.")
-		} else if len(uris) > 0 {
-			scan.results.AttributePrinterURI = string(uris[0])
+
+	var attrs []*Attribute
+	for _, g := range groups {
+		attrs = append(attrs, g.Attrs...)
+	}
+	return &attrs
+}
+
+func (scan *scan) tryReadAttributes(body string) {
+	bodyBytes := []byte(body)
+	if bytes.Contains(bodyBytes, []byte(AttributesCharset)) {
+		scan.results.Attributes = append(scan.results.Attributes, *readAllAttributes(&bodyBytes)...)
+
+		for _, attr := range scan.results.Attributes {
+			// TODO: Make this record all CUPS versions given. Currently records first version from first attribute.
```

Comment:
If having multiple CUPS versions is normal, then that would be a good idea -- but if it's a weird special case, then you are probably fine just including it once.